### ISSUE TITLE
Optimize Memory Efficiency in Adaptive Model Architecture

### ISSUE THREAD

**Issue by csmangum** [Dooders/AgentMeaning#6]
The current implementation of our adaptive models (particularly `AdaptiveEntropyBottleneck` and `FeatureGroupedVAE`) contains parameter inefficiencies that result in unnecessary memory usage, especially at certain compression levels. This causes slower training, higher memory consumption, and potentially diminished performance on resource-constrained environments.

### 1. Redundant Projection Layers
In `AdaptiveEntropyBottleneck`, we create full projection layers (`proj_down` and `proj_up`) regardless of compression level. When compression_level is close to 1.0, these become:
- Nearly identity transformations (inefficient computation)
- Unnecessary parameter overhead (memory waste)

```python
# Current implementation (always creates these regardless of compression)
self.proj_down = nn.Linear(latent_dim, self.effective_dim)
self.proj_up = nn.Linear(self.effective_dim, latent_dim * 2)
```

### 2. Excessive Parameter Scaling
For `proj_up`, we're creating a parameter matrix that scales as:
- Size: effective_dim × (latent_dim × 2)
- Example: With latent_dim=256, compression=2, this becomes 128×512 = 65,536 parameters

### 3. Multiple Full Bottlenecks in FeatureGroupedVAE
Each feature group gets its own complete bottleneck instance:
```python
self.bottlenecks[name] = AdaptiveEntropyBottleneck(
    latent_dim=group_latent_dim,
    compression_level=effective_compression,
    seed=seed,
)
```
This linearly multiplies parameter count by number of groups, despite potential for sharing.

### 1. Conditional Computation Architecture
Implement a more parameter-efficient approach that only activates necessary components:

```python
def __init__(self, latent_dim, compression_level=1.0, threshold=1.2):
    # Only create projection layers if compression exceeds threshold
    self.use_projection = compression_level >= threshold
    
    if self.use_projection:
        self.proj_down = nn.Linear(latent_dim, self.effective_dim)
        self.proj_up = nn.Linear(self.effective_dim, latent_dim * 2)
    else:
        # Use simpler parameterization for near-identity cases
        self.compress_params = nn.Parameter(torch.zeros(latent_dim, 2))
        
def forward(self, z):
    if self.use_projection:
        # Full projection path
        z_down = self.proj_down(z)
        # ... rest of current implementation
    else:
        # Simplified path for low compression
        params = self.compress_params
        mu, log_scale = params[:, 0], params[:, 1]
        # ... simplified compression logic
```

### 2. Parameter Sharing in FeatureGroupedVAE
Redesign to share components across feature groups:

```python
# Create a single shared encoder network
self.shared_compressor = nn.Module()
self.shared_compressor.mu_network = nn.Linear(latent_dim, latent_dim)
self.shared_compressor.scale_network = nn.Linear(latent_dim, latent_dim)

# Each group only needs minimal parameters
for name in group_names:
    self.group_params[name] = nn.Parameter(torch.zeros(group_latent_dim, 2))
```

### 3. Low-Rank Approximations for Large Projections
For larger projection matrices, use low-rank approximations:

```python
# Instead of single large projection
self.proj_up = nn.Sequential(
    nn.Linear(self.effective_dim, latent_dim//4),
    nn.LeakyReLU(),
    nn.Linear(latent_dim//4, latent_dim * 2)
)
```

## Expected Benefits
- **Reduced Memory Usage**: Potentially 50-80% reduction in parameter count for adaptive models
- **Faster Training**: Fewer parameters = faster gradient updates
- **More Efficient Inference**: Especially important for deployment scenarios
- **Better Scaling**: Architecture will adapt more efficiently across compression levels

## Implementation Steps
1. Profile existing model to establish parameter count baseline
2. Implement conditional computation architecture
3. Add parameter sharing for FeatureGroupedVAE
4. Implement low-rank approximations for large projections
5. Compare performance (accuracy and memory) across different compression levels
6. Update documentation to reflect new architecture

## Related Components
- `AdaptiveEntropyBottleneck`
- `FeatureGroupedVAE`
- `AdaptiveMeaningVAE`

Priority: Medium
Labels: performance, optimization, memory-efficiency
```
